File: reference-apps/nodejs/src/routes/messaging.js
const express = require('express');
const amqp = require('amqplib');
const { vaultClient } = require('../services/vault');
const config = require('../config');

const router = express.Router();

router.post('/publish/:queue', async (req, res) => {
  let connection;
  try {
    const creds = await vaultClient.getSecret('rabbitmq');
    const url = `amqp://${creds.user}:${creds.password}@${config.rabbitmq.host}:${config.rabbitmq.port}`;

    connection = await amqp.connect(url);
    const channel = await connection.createChannel();
    await channel.assertQueue(req.params.queue, { durable: true });
    channel.sendToQueue(req.params.queue, Buffer.from(JSON.stringify(req.body)));

    await channel.close();
    res.json({ queue: req.params.queue, message: req.body, published: true });
  } catch (error) {
    res.status(503).json({ error: error.message });
  } finally {
    if (connection) await connection.close().catch(() => {});
  }
});

router.get('/queue/:queue_name/info', async (req, res) => {
  let connection;
  try {
    const creds = await vaultClient.getSecret('rabbitmq');
    const url = `amqp://${creds.user}:${creds.password}@${config.rabbitmq.host}:${config.rabbitmq.port}`;

    connection = await amqp.connect(url);
    const channel = await connection.createChannel();

    // Assert queue to get its info
    const queueInfo = await channel.checkQueue(req.params.queue_name);

    await channel.close();

    res.json({
      queue: req.params.queue_name,
      message_count: queueInfo.messageCount,
      consumer_count: queueInfo.consumerCount,
      exists: true
    });
  } catch (error) {
    // Queue might not exist
    if (error.message.includes('NOT_FOUND')) {
      res.status(404).json({
        queue: req.params.queue_name,
        exists: false,
        error: 'Queue not found'
      });
    } else {
      res.status(503).json({
        queue: req.params.queue_name,
        error: error.message
      });
    }
  } finally {
    if (connection) await connection.close().catch(() => {});
  }
});

module.exports = router;

File: reference-apps/golang/internal/handlers/messaging.go
package handlers

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	amqp "github.com/rabbitmq/amqp091-go"

	"github.com/normbrandinger/devstack-core/reference-apps/golang/internal/config"
	"github.com/normbrandinger/devstack-core/reference-apps/golang/internal/services"
)

// MessagingHandler handles messaging-related endpoints
type MessagingHandler struct {
	cfg         *config.Config
	vaultClient *services.VaultClient
}

// NewMessagingHandler creates a new messaging handler
func NewMessagingHandler(cfg *config.Config, vaultClient *services.VaultClient) *MessagingHandler {
	return &MessagingHandler{
		cfg:         cfg,
		vaultClient: vaultClient,
	}
}

func (mh *MessagingHandler) getConnection(ctx context.Context) (*amqp.Connection, error) {
	creds, err := mh.vaultClient.GetSecret(ctx, "reference-api/rabbitmq")
	if err != nil {
		return nil, fmt.Errorf("failed to get Vault credentials: %w", err)
	}

	user, _ := creds["user"].(string)
	password, _ := creds["password"].(string)

	url := fmt.Sprintf("amqp://%s:%s@%s:%s/",
		user, password, mh.cfg.RabbitMQHost, mh.cfg.RabbitMQPort)

	return amqp.Dial(url)
}

// PublishMessage publishes a message to a queue
func (mh *MessagingHandler) PublishMessage(c *gin.Context) {
	queueName := c.Param("queue")

	var message map[string]interface{}
	if err := c.ShouldBindJSON(&message); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "invalid JSON payload",
		})
		return
	}

	ctx, cancel := context.WithTimeout(c.Request.Context(), 10*time.Second)
	defer cancel()

	conn, err := mh.getConnection(ctx)
	if err != nil {
		c.JSON(http.StatusServiceUnavailable, gin.H{
			"error": err.Error(),
		})
		return
	}
	defer conn.Close()

	ch, err := conn.Channel()
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": err.Error(),
		})
		return
	}
	defer ch.Close()

	// Declare queue
	q, err := ch.QueueDeclare(
		queueName,
		true,  // durable
		false, // delete when unused
		false, // exclusive
		false, // no-wait
		nil,   // arguments
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": err.Error(),
		})
		return
	}

	// Serialize message
	body, err := json.Marshal(message)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": err.Error(),
		})
		return
	}

	// Publish message
	err = ch.PublishWithContext(
		ctx,
		"",     // exchange
		q.Name, // routing key
		false,  // mandatory
		false,  // immediate
		amqp.Publishing{
			ContentType: "application/json",
			Body:        body,
		},
	)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": err.Error(),
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"queue":   queueName,
		"message": "published successfully",
	})
}

// QueueInfo returns information about a queue
func (mh *MessagingHandler) QueueInfo(c *gin.Context) {
	queueName := c.Param("queue_name")

	ctx, cancel := context.WithTimeout(c.Request.Context(), 5*time.Second)
	defer cancel()

	conn, err := mh.getConnection(ctx)
	if err != nil {
		c.JSON(http.StatusServiceUnavailable, gin.H{
			"error": err.Error(),
		})
		return
	}
	defer conn.Close()

	ch, err := conn.Channel()
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": err.Error(),
		})
		return
	}
	defer ch.Close()

	// Inspect queue
	q, err := ch.QueueInspect(queueName)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{
			"error": fmt.Sprintf("queue %s not found", queueName),
		})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"name":      q.Name,
		"messages":  q.Messages,
		"consumers": q.Consumers,
	})
}
